File: week13/11/2/P2.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> start;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> end;

        for (auto trip : trips){
            start.push(make_pair(trip[1], trip[0]));
            end.push(make_pair(trip[2], trip[0]));
        } 
        int m = 0;
        int n = 0;
        int curpas = 0;

        while (start.size() != 0 && end.size() != 0){
            if (start.top().first < end.top().first){
                curpas += start.top().second;
                if (curpas > capacity){
                    return false;
                }
                start.pop();
            }
            else{
                curpas -= end.top().second;
                end.pop();
            }
        }
        return true;     
    }
};

File: week13/11/2/P2.py
class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        start = []
        end = []
        for trip in trips:
            start.append([trip[1], trip[0]])
            end.append([trip[2], trip[0]])
        start.sort(key=lambda x: x[0])
        end.sort(key=lambda x: x[0])
        m = 0
        n = 0
        curpas = 0
        while(m < len(start) and n < len(end)):
            if start[m][0] < end[n][0]:
                curpas += start[m][1]
                m+=1
                if curpas > capacity:
                    return False
            else:
                curpas -= end[n][1]
                n+=1
        return True